crates/turborepo-cache/src/cache_archive/create.rs

```diff
@@ -0,0 +1,182 @@
+
+        if is_compressed {
+            let zw = zstd::Encoder::new(file_buffer, 0)?;
+
+            Ok(CacheArchive::Compressed(tar::Builder::new(zw)))
+        } else {
+            Ok(CacheArchive::Uncompressed(tar::Builder::new(file_buffer)))
+        }
+    }
+
+    // Adds a user-cached item to the tar
+    fn add_file(
+        &mut self,
+        anchor: &AbsoluteSystemPath,
+        file_path: &AnchoredSystemPath,
+    ) -> Result<(), CacheError> {
+        // Resolve the fully-qualified path to the file to read it.
+        let source_path = anchor.resolve(file_path);
+
+        // Grab the file info to construct the header.
+        let file_info = source_path.symlink_metadata()?;
+
+        // Normalize the path within the cache
+        let file_path: RelativeUnixPathBuf = file_path.try_into()?;
+        let canonical_file_path = file_path.make_canonical_for_tar(file_info.is_dir());
+
+        let mut header = Self::create_header(&source_path, &file_info)?;
+
+        if matches!(header.entry_type(), EntryType::Regular) && file_info.len() > 0 {
+            let file = source_path.open()?;
+            self.append_data(&mut header, canonical_file_path.as_path()?, file)?;
+        } else {
+            self.append_data(
+                &mut header,
+                canonical_file_path.as_path()?,
+                &mut std::io::empty(),
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn create_header(
+        source_path: &AbsoluteSystemPath,
+        file_info: &fs::Metadata,
+    ) -> Result<Header, CacheError> {
+        let mut header = Header::new_gnu();
+
+        let mode: u32;
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::MetadataExt;
+            mode = file_info.mode();
+        }
+        #[cfg(windows)]
+        {
+            // Windows makes up 0o666 for files, which in the Go code
+            // we do: (0o666 & 0o755) | 0o111 which produces 0o755
+            mode = 0o755
+        }
+        header.set_mode(mode);
+
+        // Do we need to populate the additional linkname field in Header?
+        if file_info.is_symlink() {
+            let link = source_path.read_link()?;
+            header.set_link_name(link)?;
+        }
+
+        // Throw an error if trying to create a cache that contains a type we don't
+        // support.
+        if !matches!(
+            header.entry_type(),
+            EntryType::Regular | EntryType::Directory | EntryType::Symlink
+        ) {
+            return Err(CacheError::UnsupportedFileType(
+                header.entry_type(),
+                Backtrace::capture(),
+            ));
+        }
+
+        // Consistent creation
+        header.set_uid(0);
+        header.set_gid(0);
+        header.as_gnu_mut().unwrap().set_atime(0);
+        header.set_mtime(0);
+        header.as_gnu_mut().unwrap().set_ctime(0);
+
+        Ok(header)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Sorry mistyped, I meant `cacheitem/create_test.go`

Reply:
Could we do this in a follow up PR? There's a few things I need to do around getting a SHA that will expand this PR

Reply:
I think it would be good to have those available. You could do it in a separate stacked PR if you don't want to pollute this one, but that's the real assurance that this code is correct.

Reply:
I can open a stacked PR, but it'd be nice to merge this one before opening another PR